core/filterconstraint.cpp

```diff
@@ -0,0 +1,939 @@
+{
+	return true;
+}
+*/
+
+static bool has_suits(const filter_constraint &c, const struct dive *d)
+{
+	QStringList diveSuits;
+	if (d->suit)
+		diveSuits.push_back(QString(d->suit));
+	return check(c, diveSuits);
+}
+
+static bool has_notes(const filter_constraint &c, const struct dive *d)
+{
+	QStringList diveNotes;
+	if (d->notes)
+		diveNotes.push_back(QString(d->notes));
+	return check(c, diveNotes);
+}
+
+static bool check_numerical_range(const filter_constraint &c, int v)
+{
+	switch (c.range_mode) {
+	case FILTER_CONSTRAINT_EQUAL:
+		return (v == c.data.numerical_range.from) != c.negate;
+	case FILTER_CONSTRAINT_LESS:
+		return (v <= c.data.numerical_range.to) != c.negate;
+	case FILTER_CONSTRAINT_GREATER:
+		return (v >= c.data.numerical_range.from) != c.negate;
+	case FILTER_CONSTRAINT_RANGE:
+		return (v >= c.data.numerical_range.from && v <= c.data.numerical_range.to) != c.negate;
+	}
+	return false;
+}
+
+// Check a numerical range, but consider a value of 0 as "not set".
+static bool check_numerical_range_non_zero(const filter_constraint &c, int v)
+{
+	if (v == 0)
+		return c.negate;
+	return check_numerical_range(c, v);
+}
+
+static long days_since_epoch(timestamp_t timestamp)
+{
+	return timestamp / (3600 * 24);
+}
+
+static int seconds_since_midnight(timestamp_t timestamp)
+{
+	return timestamp % (3600 * 24);
+}
+
+static bool check_date_range(const filter_constraint &c, const struct dive *d)
+{
+	// We don't consider dives past midnight. Should we?
+	long day = days_since_epoch(d->when);
+	switch (c.range_mode) {
+	case FILTER_CONSTRAINT_EQUAL:
+		return (day == days_since_epoch(c.data.timestamp_range.from)) != c.negate;
+	case FILTER_CONSTRAINT_LESS:
+		return (day <= days_since_epoch(c.data.timestamp_range.to)) != c.negate;
+	case FILTER_CONSTRAINT_GREATER:
+		return (day >= days_since_epoch(c.data.timestamp_range.from)) != c.negate;
+	case FILTER_CONSTRAINT_RANGE:
+		return (day >= days_since_epoch(c.data.timestamp_range.from) &&
+			day <= days_since_epoch(c.data.timestamp_range.to)) != c.negate;
+	}
+	return false;
+}
+
+static bool check_datetime_range(const filter_constraint &c, const struct dive *d)
+{
+	switch (c.range_mode) {
+	case FILTER_CONSTRAINT_EQUAL:
+		// Exact mode is a bit strange for timestamps. Therefore we return any dive
+		// where the given timestamp is during that dive.
+		return time_during_dive_with_offset(d, c.data.timestamp_range.from, 0) != c.negate;
+	case FILTER_CONSTRAINT_LESS:
+		return (dive_endtime(d) <= c.data.timestamp_range.to) != c.negate;
+	case FILTER_CONSTRAINT_GREATER:
+		return (d->when >= c.data.timestamp_range.from) != c.negate;
+	case FILTER_CONSTRAINT_RANGE:
+		return (d->when >= c.data.timestamp_range.from &&
+			dive_endtime(d) <= c.data.timestamp_range.to) != c.negate;
+	}
+	return false;
+}
+
+static bool check_time_of_day_internal(const dive *d, enum filter_constraint_range_mode range_mode, int from, int to, bool negate)
```

Comment:
so sometimes you pass in the constraint, sometimes you pass in just a few elements. Why?

Reply:
This is a special case that is explained in the comment in the next function. I added a small comment also here.